node-wasm/src/ports.rs

```diff
@@ -0,0 +1,306 @@
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
+        forward_messages_to: mpsc::UnboundedSender<(
+            ClientId,
+            Result<NodeCommand, TypedMessagePortError>,
+        )>,
+        forward_connects_to: mpsc::UnboundedSender<JsValue>,
+    ) -> Result<Self> {
+        let onmessage = Closure::new(move |ev: MessageEvent| {
+            let message_tx = forward_messages_to.clone();
+            let port_tx = forward_connects_to.clone();
+            spawn_local(async move {
+                let message: Result<NodeCommand, _> =
+                    from_value(ev.data()).map_err(TypedMessagePortError::FailedToConvertValue);
+
+                let ports = ev.ports();
+                if Array::is_array(&ports) {
+                    let port = ports.get(0);
+                    if !port.is_undefined() {
+                        if let Err(e) = port_tx.send(port) {
+                            error!("port forwarding channel closed, shouldn't happen: {e}");
+                        }
+                    }
+                }
+
+                if let Err(e) = message_tx.send((id, message)) {
+                    error!("message forwarding channel closed, shouldn't happen: {e}");
+                }
+            })
+        });
+
+        let port = prepare_message_port(object, &onmessage)
+            .context("failed to setup port for ClientConnection")?;
+
+        Ok(ClientConnection {
+            port,
+            _onmessage: onmessage,
+        })
+    }
+
+    fn send(&self, message: &WorkerResponse) -> Result<()> {
+        let serializer = Serializer::json_compatible();
+        let message_value = message
+            .serialize(&serializer)
+            .context("could not serialise message")?;
+        self.port
+            .post_message(&message_value)
+            .context("could not send command to worker")?;
+        Ok(())
+    }
+}
+
+pub struct RequestServer {
+    ports: Vec<ClientConnection>,
+    connect_tx: mpsc::UnboundedSender<JsValue>,
+    connect_rx: mpsc::UnboundedReceiver<JsValue>,
+    request_tx: mpsc::UnboundedSender<(ClientId, Result<NodeCommand, TypedMessagePortError>)>,
+    request_rx: mpsc::UnboundedReceiver<(ClientId, Result<NodeCommand, TypedMessagePortError>)>,
+}
+
+impl RequestServer {
+    pub fn new() -> RequestServer {
+        let (request_tx, request_rx) = mpsc::unbounded_channel();
+        let (connect_tx, connect_rx) = mpsc::unbounded_channel();
+
+        RequestServer {
+            ports: vec![],
+            connect_tx,
+            connect_rx,
+            request_tx,
+            request_rx,
+        }
+    }
+
+    pub async fn recv(&mut self) -> (ClientId, Result<NodeCommand, TypedMessagePortError>) {
```

Comment:
I think we should just return `(ClientId, NodeCommand)` here. At the call site we just do:
```rs
        let (client_id, result) = server.recv().await;

        let command = result
            .with_context(|| format!("could not parse command received from {client_id:?}"))?;
```
so we could move the error reporting here and simplify types around

Reply:
Moved error logging into callback and skipped the error passing via channels as suggested :+1: